src/network/tcpsocket.c

```diff
@@ -80,50 +80,48 @@ static void client_to_server_state(struct ustream *s) {
 }
 
 static void client_read_cb(struct ustream *s, int bytes) {
-    char *str;
+    char *str, *str_tmp;
     int len = 0;
-    uint32_t final_len;
-    int max_retry = 3, tried = 0;
-
-    do {
-        str = ustream_get_read_buf(s, &len);
-        if (!str)
-            break;
-
-        if (network_config.use_symm_enc) {
-            final_len = ntohl(*(uint32_t *)str);
-            if(len < final_len) {//not complete msg, wait for next recv
-                fprintf(stderr,"not complete msg, len:%d, expected len:%u\n", len, final_len);
-                if (tried++ == max_retry) {
-                    ustream_consume(s, len);
-                    return;//drop package
-                }
-                continue;
-            }
-            char *dec = gcrypt_decrypt_msg(str+sizeof(final_len), final_len-sizeof(final_len));
-        
-            handle_network_msg(dec);
-            free(dec);
-            ustream_consume(s, final_len);//one msg is processed
-            tried = 0;
-        } else {
-            final_len = ntohl(*(uint32_t *)str);
-            if(len < final_len){
-                fprintf(stderr,"not complete msg, len:%d, expected len:%u\n", len, final_len);
-                if (tried++ == max_retry) {
-                    ustream_consume(s, len);
-                    return;
-                }
-                continue;
-            } 
-            char* msg = malloc(final_len);
-            memcpy(msg, str+sizeof(final_len), final_len-sizeof(final_len));
-            handle_network_msg(msg);
-            ustream_consume(s, final_len);//one msg is processed
-            free(msg);
-            tried = 0;
-        }
-    } while(1);
+    uint32_t final_len = sizeof(uint32_t);
+    str = malloc(final_len);
+    if (!str) {
+        fprintf(stderr,"not enough memory\n");
+        goto memory_full;
+    }
+
+    if ((len = ustream_read(s, str, final_len)) < final_len){//ensure recv sizeof(uint32_t).
```

Comment:
Don't we have to call ustream_pending_data first?

Reply:
i got your point, i would modify this patch.